### DISCUSSION TITLE
How to write a function returning a closure useable with middleware::from_fn?

### DISCUSSION THREAD

**Discussion by martinetd** [tokio-rs/axum#2649] [Q&A]
### Summary

This is probably more of a rust question than an axum question...

I can make middleware::from_fn work with a closure e.g. (`layer(middleware::from_fn(|r, n| do_something(r, n, "test")))`, but I can't figure how to please the rust type system in order to make a "factory" for do_something, e.g. how to make this simplified code compile:
```
use axum::{
    extract::Request,
    middleware::{self, Next},
    response::{Response},
    routing::post,
    Router,
};
use std::{future::Future, pin::Pin};

#[tokio::main]
async fn main() {
    let app = Router::new()
        .route("/", post(|| async move { "Hello from `POST /`" }))
        // this does not work:
        .layer(middleware::from_fn(make_do_something("test")));
        // this works instead:
        //.layer(middleware::from_fn(|r, n| do_something(r, n, "test")));
    
    let listener = tokio::net::TcpListener::bind("127.0.0.1:3000")
        .await
        .unwrap();
    println!("listening on {}", listener.local_addr().unwrap());
    axum::serve(listener, app).await.unwrap();
}

fn make_do_something(
    data: &'static str,
) -> impl Fn(Request, Next) -> Pin<Box<dyn Future<Output = Response>>> + Clone { 
    move |request: Request, next: Next| Box::pin(do_something(request, next, data))
}       
    
async fn do_something(
    req: Request,
    next: Next,
    data: &str, 
) -> Response {
    println!("Called with {}", data);
    next.run(req).await
}
```

It currently fails with:
```
error[E0277]: the trait bound `axum::middleware::FromFn<impl (Fn(axum::http::Request<Body>, Next) -> Pin<Box<(dyn Future<Output = Response<Body>> + 'static)>>) + Clone, (), Route, _>: tower_service::Service<axum::http::Request<Body>>` is not satisfied
   --> examples/print-request-response/src/main.rs:21:16
    |
21  |         .layer(middleware::from_fn(make_do_something("test")));
    |          ----- ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `tower_service::Service<axum::http::Request<Body>>` is not implemented for `axum::middleware::FromFn<impl (Fn(axum::http::Request<Body>, Next) -> Pin<Box<(dyn Future<Output = Response<Body>> + 'static)>>) + Clone, (), Route, _>`
    |          |
    |          required by a bound introduced by this call
    |
    = help: the following other types implement trait `tower_service::Service<Request>`:
              axum::middleware::FromFn<F, S, I, (T1,)>
              axum::middleware::FromFn<F, S, I, (T1, T2)>
              axum::middleware::FromFn<F, S, I, (T1, T2, T3)>
              axum::middleware::FromFn<F, S, I, (T1, T2, T3, T4)>
              axum::middleware::FromFn<F, S, I, (T1, T2, T3, T4, T5)>
              axum::middleware::FromFn<F, S, I, (T1, T2, T3, T4, T5, T6)>
              axum::middleware::FromFn<F, S, I, (T1, T2, T3, T4, T5, T6, T7)>
              axum::middleware::FromFn<F, S, I, (T1, T2, T3, T4, T5, T6, T7, T8)>
            and 8 others
note: required by a bound in `Router::<S>::layer`
   --> /home/martinet/code/rust/axum/axum/src/routing/mod.rs:278:21
    |
275 |     pub fn layer<L>(self, layer: L) -> Router<S>
    |            ----- required by a bound in this associated function
...
278 |         L::Service: Service<Request> + Clone + Send + 'static,
    |                     ^^^^^^^^^^^^^^^^ required by this bound in `Router::<S>::layer`

For more information about this error, try `rustc --explain E0277`.
```

I've tried to add a few more traits to the return type as well as the nightly impl_trait_in_fn_trait_return feature with "impl Future" instead of the pin box to no avail... Do you know if there's a way to do this?

Otherwise I'll stick to the explicit closure, I thought I had to be explicit about the types in there but as seen here something as short as `|r, n| foo(r, n, "something")` works so it's much shorter than my original code spelling everything out

### axum version

tip of main branch

**Comment by jplatte** [comment]
It's certainly possible. You can find the bounds your fn needs to fulfill [here](https://docs.rs/axum/latest/axum/middleware/struct.FromFn.html#impl-Service%3CRequest%3CBody%3E%3E-for-FromFn%3CF,+S,+I,+(T1,)%3E) and below (that one is the one you need though if your closure takes just one argument before `Next`).

In your specific example, the `dyn Future` is missing a `+ Send`. You don't have to specify that with `impl Future` though, so there's probably something else going on as well. Can't tell you what at a glance though.

**Reply by martinetd** [reply]
I did try adding + Send (and +'static) as well to no avail (no difference in the error message, so the requirement for that trait isn't fulfilled); I've only left the Clone because that yields another error when cloning the closure when building the FromFn and that's unrelated.
(Also noticed it needs to be FnMut and not Fn, but that didn't help either)

I'm thinking that since I have to add a `Pin<Box<>>` the type isn't the same so rust doesn't recognize it?


I've also tried adding a new "AsyncFn2" trait as described in https://github.com/rust-lang/rust/pull/93582 but that's also a new trait so I don't think it'd work unless perhaps if everything is built together... the error is pretty similar saying the trait bound isn't satisfied:
```
trait AsyncFn2<A, B>: FnMut(A, B) -> <Self as AsyncFn2<A, B>>::Future {
    type Future: Future<Output = Self::Out>;

    type Out;
} 
    
impl<A, Fut, F> AsyncFn2<A, Next> for F
where
    F: FnMut(A, Next) -> Fut + Clone + Send + 'static,
    Fut: Future + Send + 'static,
{   
    type Future = Fut;

    type Out = Fut::Output;
}

fn make_do_something(
    data: &'static str,
) -> impl AsyncFn2<Request, Next, Out = Response> + Clone + Send + 'static {
    move |request: Request, next: Next| Box::pin(do_something(request, next, data))
}     
```
```
error[E0277]: the trait bound `axum::middleware::FromFn<impl AsyncFn2<axum::http::Request<Body>, Next, Out = Response<Body>> + Clone + Send + 'static, (), Route, _>: tower_service::Service<axum::http::Request<Body>>` is not satisfied
   --> examples/print-request-response/src/main.rs:15:16
    |
15  |         .layer(middleware::from_fn(make_do_something("test")));
    |          ----- ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `tower_service::Service<axum::http::Request<Body>>` is not implemented for `FromFn<impl AsyncFn2<..., ..., Out = ...> + Clone + Send, ..., ..., ...>`
    |          |
    |          required by a bound introduced by this call
    |
    = help: the following other types implement trait `tower_service::Service<Request>`:
              axum::middleware::FromFn<F, S, I, (T1,)>
...
```

I'm not sure it's worth spending much more time on this right now, I'll just stick with explicit closures.
If someones who's done it before can point me at a working example later then I'll give it another look (and perhaps submit a PR with an example in axum's examples/ directory), but until then I can spare a couple of extra lines in ~10 call sites :)
